waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
+
+# Performs a Diffie-Hellman operation between two elliptic curve keys (one private, one public)
+proc dh*(private: EllipticCurveKey, public: EllipticCurveKey): EllipticCurveKey =
+
+  # The output result of the Diffie-Hellman operation
+  var output: EllipticCurveKey
+
+  # Since the EC multiplication writes the result to the input, we copy the input to the output variable 
+  output = public
+  # We execute the DH operation
+  EllipticCurve.mul(output, private)
+
+  return output
+
 
 #################################################################
 
+# Noise state machine primitives
+
+# Overview :
+# - Alice and Bob process (i.e. read and write, based on their role) each token appearing in a handshake pattern, consisting of pre-message and message patterns;
+# - Both users initialize and update according to processed tokens a Handshake State, a Symmetric State and a Cipher State;
+# - A preshared key psk is processed by calling MixKeyAndHash(psk);
+# - When an ephemeral public key e is read or written, the handshake hash value h is updated by calling mixHash(e); If the handshake expects a psk, MixKey(e) is further called
+# - When an encrypted static public key s or a payload message m is read, it is decrypted with decryptAndHash;
+# - When a static public key s or a payload message is writted, it is encrypted with encryptAndHash;
+# - When any Diffie-Hellman token ee, es, se, ss is read or written, the chaining key ck is updated by calling MixKey on the computed secret;
+# - If all tokens are processed, users compute two new Cipher States by calling Split;
+# - The two Cipher States obtained from Split are used to encrypt/decrypt outbound/inbound messages.
+
+#################################
+# Cipher State Primitives
+#################################
+
+# Checks if a Cipher State has an encryption key set
+proc hasKey(cs: CipherState): bool =
+  return (cs.k != EmptyKey)
+
+# Encrypts a plaintext using key material in a Noise Cipher State
+# The CipherState is updated increasing the nonce (used as a counter in Noise) by one
+proc encryptWithAd*(state: var CipherState, ad, plaintext: openArray[byte]): seq[byte]
+  {.raises: [Defect, NoiseNonceMaxError].} =
+  
+  # The output is the concatenation of the ciphertext and authorization tag
+  # We define its length accordingly
+  var ciphertext = newSeqOfCap[byte](plaintext.len + sizeof(ChaChaPolyTag))
+  
+  # Since ChaChaPoly encryption primitive overwrites the input with the output,
+  # we copy the plaintext in the output ciphertext variable and we pass it to encryption
+  ciphertext.add(plaintext)
+
+  # The nonce is read from the input CipherState
+  # By Noise specification the nonce is 4 bytes long out of the 12 bytes supported by ChaChaPoly
+  var nonce: ChaChaPolyNonce
+  nonce[4..<12] = toBytesLE(state.n)
+
+  # We perform encryption and we store the authorization tag
+  var authorizationTag: ChaChaPolyTag
+  ChaChaPoly.encrypt(state.k, nonce, authorizationTag, ciphertext, ad)
```

Comment:
This is correct and it should be checked here as well rather than only in encryptAndHash/decryptAndHash. Since the latter always call encryptWithAd/decryptWithAd, respective, I moved the check there.

Reply:
Addressed in https://github.com/status-im/nwaku/pull/933/commits/4ae4364845643190a3af13b3f15fd4afb6bcf906